htdp-lib/2htdp/abstraction.rkt

```diff
@@ -114,7 +115,33 @@
     [(_ type:id case:expr [variant:id (field:id ...) body:expr] ...)
      #'(type-case type case [variant (field ...) body] ...)]))
 
+(begin-for-syntax
+  (define-syntax-class pattern-cls
+    #:description "pattern"
+    #:attributes (expr)
+    (pattern name:id
+             #:with expr #'name)
+
+    ;; literal-constants
+    (pattern {~or* :string :number :boolean :char :regexp :bytes :byte-regexp}
+             #:with expr this-syntax)
+    (pattern ({~datum quote} x)
```

Comment:
*SL programmers could technically import with renaming, and get `foo:cons` etc going. I think ignoring this for now and using `~datum` is okay. (I think.)